Add panda fw and dongle id to params

martinl opened the pull request:
This PR gets Panda firmware version and Panda dongle ID and adds them to params. I will submit separate PR to offroad apk to show them under Developer menu

martinl commented:
openpilot-apks PR https://github.com/commaai/openpilot-apks/pull/12

geohot commented:
Can you fix the CI?

geohot commented:
Hmm, is this tested?

I was about to merge your apk PR, but then I read this code. Is Params imported?

martinl commented:
Fixed the CI and set up local docker for running tests before next PR. Replugging and swapping pandas is a good point, I'll reimplement setting the panda params in boardd

geohot commented:
When that's done, I'll merge.

rbiasini commented:
Or you could add the rule `CLEAR_ON_PANDA_DISCONNECT`.

geohot commented:
That will clear them, but it won't readd them for the new panda.

martinl commented:
Implemented reading the panda firmware and serial values in boardd. Also fixed permissions so that offroad app can read params written by write_db_value(). Tested and working. Please review and comment

martinl commented:
rebased the branch to 0.6.3 devel

rbiasini commented on selfdrive/pandad.py:
```diff
@@ -65,9 +65,8 @@ def update_panda():
 
-
 def main(gctx=None):
-  update_panda()
```
probably a leftover from tests?

rbiasini commented on selfdrive/common/params.cc:
```diff
@@ -78,6 +79,12 @@ int write_db_value(const char* params_path, const char* key, const char* value,
+  result = fchmod(tmp_fd, 0644);
+  if (result < 0) {
+    goto cleanup;
+  }
```
not needed, is it?

martinl replied:
chmod 0644 is needed for offroad app to read the params file

rbiasini replied:
ah, right. Haven't gone through tha APK/display part yet. Short term I'm fine making sure those params are logged, but yeah, would be nice to display them in developer settings.

rbiasini commented:
Thanks for the PR. Due to conflicts and some remaining issues in the PR, I ended up merging the following diff in our internal repo (will be in next openpilot release).
@martinl , let me know if you spot anything wrong.
```
diff --git a/common/params.py b/common/params.py
index 3c5a63786..7f31e07bb 100755
--- a/common/params.py
+++ b/common/params.py
@@ -77,6 +77,8 @@ class UnknownKeyName(Exception):
   "LiveParameters": [TxType.PERSISTENT],
   "LongitudinalControl": [TxType.PERSISTENT],
   "OpenpilotEnabledToggle": [TxType.PERSISTENT],
+  "PandaFirmware": [TxType.CLEAR_ON_MANAGER_START, TxType.CLEAR_ON_PANDA_DISCONNECT],
+  "PandaDongleId": [TxType.CLEAR_ON_MANAGER_START, TxType.CLEAR_ON_PANDA_DISCONNECT],
   "Passive": [TxType.PERSISTENT],
   "RecordFront": [TxType.PERSISTENT],
   "ReleaseNotes": [TxType.PERSISTENT],
diff --git a/selfdrive/boardd/boardd.cc b/selfdrive/boardd/boardd.cc
index ee11d210d..f7f5ab277 100644
--- a/selfdrive/boardd/boardd.cc
+++ b/selfdrive/boardd/boardd.cc
@@ -123,6 +123,13 @@ void *safety_setter_thread(void *s) {
 bool usb_connect() {
   int err;
   unsigned char hw_query[1] = {0};
+  unsigned char fw_ver_buf[64];
+  unsigned char serial_buf[16];
+  const char *fw_ver;
+  const char *serial;
+  int fw_ver_sz = 0;
+  int serial_sz = 0;
+
   ignition_last = false;
 
   dev_handle = libusb_open_device_with_vid_pid(ctx, 0xbbaa, 0xddcc);
@@ -138,6 +145,27 @@ bool usb_connect() {
     libusb_control_transfer(dev_handle, 0xc0, 0xe5, 1, 0, NULL, 0, TIMEOUT);
   }
 
+  // get panda fw
+  err = libusb_control_transfer(dev_handle, 0xc0, 0xd6, 0, 0, fw_ver_buf, 64, TIMEOUT);
+  if (err > 0) {
+    fw_ver = (const char *)fw_ver_buf;
+    fw_ver_sz = err;
+    write_db_value(NULL, "PandaFirmware", fw_ver, fw_ver_sz);
+    printf("panda fw: %.*s\n", fw_ver_sz, fw_ver);
+  }
+  else { goto fail; }
+
+  // get panda serial
+  err = libusb_control_transfer(dev_handle, 0xc0, 0xd0, 0, 0, serial_buf, 16, TIMEOUT);
+
+  if (err > 0) {
+    serial = (const char *)serial_buf;
+    serial_sz = err;
+    write_db_value(NULL, "PandaDongleId", serial, serial_sz);
+    printf("panda serial: %.*s\n", serial_sz, serial);
+  }
+  else { goto fail; }
+
   // power off ESP
   libusb_control_transfer(dev_handle, 0xc0, 0xd9, 0, 0, NULL, 0, TIMEOUT);
 ```
